File: server-shared/src/test/java/eu/darken/porter/endpoint/PorterEndpointPermissionStateTest.java
package eu.darken.porter.endpoint;

import static eu.darken.porter.protocol.PorterProtocol.ATTACH_PACKAGE_NAME;
import static eu.darken.porter.protocol.PorterProtocol.ATTACH_PROTOCOL_VERSION;
import static eu.darken.porter.protocol.PorterProtocol.REPLY_PERMISSION_GRANTED;
import static eu.darken.porter.protocol.PorterProtocol.REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static rikka.shizuku.server.ServerTestSupport.newCore;

import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;
import org.robolectric.shadows.ShadowBinder;

import java.util.Collections;
import java.util.List;

import eu.darken.porter.core.ManagerOperations;
import eu.darken.porter.protocol.PorterProtocol;
import eu.darken.porter.server.IPorterApplication;
import rikka.shizuku.server.ClientManager;
import rikka.shizuku.server.ClientRecord;
import rikka.shizuku.server.ConfigManager;
import rikka.shizuku.server.ServerTestSupport.TestPolicy;
import rikka.shizuku.server.ServerTestSupport.TestUserServiceManager;
import rikka.shizuku.server.util.HandlerUtil;

/** What a Porter client is told when the grant behind its record changes after it attached. */
@RunWith(RobolectricTestRunner.class)
@Config(sdk = 34, manifest = Config.NONE)
public class PorterEndpointPermissionStateTest {

    private static final String PACKAGE = "eu.darken.porter.probe";
    /** User 10, app id 10200, so the user id derived from it is not the trivial zero. */
    private static final int CLIENT_UID = 1010200;
    private static final int CLIENT_PID = 45678;

    private ConfigManager config;
    private ClientManager<ConfigManager> clients;
    private PorterEndpoint endpoint;

    @Before
    public void setup() {
        HandlerUtil.setMainHandler(mock(Handler.class));
        config = mock(ConfigManager.class);
        clients = new ClientManager<>(config);
        endpoint = new PorterEndpoint(
                newCore(clients, new TestUserServiceManager(), config, new TestPolicy(),
                        uid -> Collections.singletonList(PACKAGE)),
                mock(ManagerOperations.class));

        ShadowBinder.setCallingUid(CLIENT_UID);
        ShadowBinder.setCallingPid(CLIENT_PID);
    }

    @After
    public void teardown() {
        ShadowBinder.reset();
    }

    private static IPorterApplication porterApplication(IBinder binder) {
        IPorterApplication application = mock(IPorterApplication.class);
        when(application.asBinder()).thenReturn(binder);
        return application;
    }

    private static Bundle attachArgs(String packageName) {
        Bundle args = new Bundle();
        args.putString(ATTACH_PACKAGE_NAME, packageName);
        args.putInt(ATTACH_PROTOCOL_VERSION, PorterProtocol.VERSION);
        return args;
    }

    @Test
    public void anAttachedClientIsSentBothFlagsOfEveryStateChange() throws Exception {
        IPorterApplication application = porterApplication(mock(IBinder.class));
        endpoint.attach(application, attachArgs(PACKAGE));

        ClientRecord record = clients.findClient(CLIENT_UID, CLIENT_PID);
        assertNotNull(record);

        record.callback.onPermissionStateChanged(false, true);
        record.callback.onPermissionStateChanged(true, false);

        ArgumentCaptor<Bundle> state = ArgumentCaptor.forClass(Bundle.class);
        verify(application, times(2)).dispatchPermissionStateChanged(state.capture());

        List<Bundle> sent = state.getAllValues();
        assertEquals(2, sent.size());

        assertFalse(sent.get(0).getBoolean(REPLY_PERMISSION_GRANTED));
        assertTrue(sent.get(0).getBoolean(REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE));

        assertTrue(sent.get(1).getBoolean(REPLY_PERMISSION_GRANTED));
        assertFalse(sent.get(1).getBoolean(REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE));
    }
}

File: server-shared/src/test/java/eu/darken/porter/endpoint/PorterEndpointPermissionStateTest.kt
package eu.darken.porter.endpoint

import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import eu.darken.porter.protocol.PorterProtocol
import eu.darken.porter.protocol.PorterProtocol.ATTACH_PACKAGE_NAME
import eu.darken.porter.protocol.PorterProtocol.ATTACH_PROTOCOL_VERSION
import eu.darken.porter.protocol.PorterProtocol.REPLY_PERMISSION_GRANTED
import eu.darken.porter.protocol.PorterProtocol.REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE
import eu.darken.porter.server.IPorterApplication
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.Mockito.mock
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowBinder
import rikka.shizuku.server.ClientManager
import rikka.shizuku.server.ConfigManager
import rikka.shizuku.server.ServerTestSupport
import rikka.shizuku.server.util.HandlerUtil

/** What a Porter client is told when the grant behind its record changes after it attached. */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34], manifest = Config.NONE)
class PorterEndpointPermissionStateTest {

    private lateinit var config: ConfigManager
    private lateinit var clients: ClientManager<ConfigManager>
    private lateinit var endpoint: PorterEndpoint

    @Before
    fun setup() {
        HandlerUtil.mainHandler = mock(Handler::class.java)
        config = mock(ConfigManager::class.java)
        clients = ClientManager(config)
        endpoint = PorterEndpoint(
            ServerTestSupport.newCore(clients, ServerTestSupport.TestUserServiceManager(), config, ServerTestSupport.TestPolicy()) {
                listOf(PACKAGE)
            },
        )

        ShadowBinder.setCallingUid(CLIENT_UID)
        ShadowBinder.setCallingPid(CLIENT_PID)
    }

    @After
    fun teardown() {
        ShadowBinder.reset()
    }

    @Test
    @Throws(Exception::class)
    fun anAttachedClientIsSentBothFlagsOfEveryStateChange() {
        val application = porterApplication(mock(IBinder::class.java))
        endpoint.attach(application, attachArgs(PACKAGE))

        val record = clients.findClient(CLIENT_UID, CLIENT_PID)
        assertNotNull(record)

        record!!.callback.onPermissionStateChanged(false, true)
        record.callback.onPermissionStateChanged(true, false)

        val state = ArgumentCaptor.forClass(Bundle::class.java)
        verify(application, times(2)).dispatchPermissionStateChanged(state.capture())

        val sent = state.allValues
        assertEquals(2, sent.size)

        assertFalse(sent[0].getBoolean(REPLY_PERMISSION_GRANTED))
        assertTrue(sent[0].getBoolean(REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE))

        assertTrue(sent[1].getBoolean(REPLY_PERMISSION_GRANTED))
        assertFalse(sent[1].getBoolean(REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE))
    }

    private companion object {
        const val PACKAGE = "eu.darken.porter.probe"

        /** User 10, app id 10200, so the user id derived from it is not the trivial zero. */
        const val CLIENT_UID = 1010200
        const val CLIENT_PID = 45678

        fun porterApplication(binder: IBinder): IPorterApplication {
            val application = mock(IPorterApplication::class.java)
            `when`(application.asBinder()).thenReturn(binder)
            return application
        }

        fun attachArgs(packageName: String): Bundle = Bundle().apply {
            putString(ATTACH_PACKAGE_NAME, packageName)
            putInt(ATTACH_PROTOCOL_VERSION, PorterProtocol.VERSION)
        }
    }
}
